File: BOJ/1260_1.py
'''
    DFS와 BFS
    https://www.acmicpc.net/problem/1260
'''
import sys
import queue
sys.setrecursionlimit(1000000)
input = sys.stdin.readline
N, M, V = map(int, input().split())
pick = []
visited = [0] * (N + 1)
Graph = [[] * (M + 1) for _ in range(N+1)]
for _ in range(M):
    u, v = map(int, input().split())
    Graph[u].append(v)
    Graph[v].append(u)
for i in range(1, N + 1):
    Graph[i].sort()


def dfs(node):
    visited[node] = True
    pick.append(node)
    for i in range(len(Graph[node])):
        next = Graph[node][i]
        if(visited[next] == False):
            dfs(next)


def bfs(start):
    q = queue.Queue()
    q.put(start)
    visited[start] = True
    pick.append(start)

    while q.empty() == False:
        current = q.get()
        for i in range(len(Graph[current])):
            next = Graph[current][i]
            if(visited[next] == False):
                q.put(next)
                visited[next] = True
                pick.append(next)


dfs(V)
for node in pick:
    print(node, end=" ")
print()
pick.clear()
visited = [0] * (N + 1)
bfs(V)

for node in pick:
    print(node, end=" ")

'''
    참고
    큐 사용법
    https://m.blog.naver.com/PostView.nhn?blogId=dudwo567890&logNo=130165707606&proxyReferer=https%3A%2F%2Fwww.google.com%2F
'''

File: BOJ/1260_1.cpp
// DFS와 BFS
// https://www.acmicpc.net/problem/1260
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;
int N, M, V;
const int MAX = 1000 + 1;
vector<int> graph[MAX], picked;
bool visited[MAX];
void bfs(int start)
{
    queue<int> q;
    q.push(start);
    visited[start] = true;
    picked.push_back(start);
    while (!q.empty())
    {
        int current = q.front();
        q.pop();
        for (int i = 0; i < graph[current].size(); i++)
        {
            int next = graph[current][i];
            if (!visited[next])
            {
                q.push(next);
                visited[next] = true;
                picked.push_back(next);
            }
        }
    }
}
void dfs(int current)
{
    visited[current] = true;
    picked.push_back(current);
    for (int i = 0; i < graph[current].size(); i++)
    {
        int next = graph[current][i];
        if (!visited[next])
            dfs(next);
    }
}
void init()
{
    cin >> N >> M >> V;
    int u, v;
    for (int i = 0; i < M; i++)
    {
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    // 정점번호가 작은 것 순으로 정렬
    for (int i = 1; i <= N; i++)
        sort(graph[i].begin(), graph[i].end());
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    init();
    dfs(V);
    vector<int>::iterator iter;
    for (iter = picked.begin(); iter != picked.end(); iter++)
        cout << *iter << ' ';
    picked.clear();
    cout << '\n';
    fill_n(visited, MAX, 0);
    bfs(V);
    for (iter = picked.begin(); iter != picked.end(); iter++)
        cout << *iter << ' ';
    return 0;
}